bench-tps/src/bench_tps_client/high_tps_client.rs

```diff
@@ -0,0 +1,233 @@
+    solana_client::{
+        connection_cache::Protocol, nonblocking::tpu_client::LeaderTpuService,
+        tpu_connection::TpuConnection,
+    },
+    solana_connection_cache::connection_cache::ConnectionCache as BackendConnectionCache,
+    solana_quic_client::{QuicConfig, QuicConnectionManager, QuicPool},
+    solana_rpc_client::rpc_client::RpcClient,
+    solana_rpc_client_api::config::RpcBlockConfig,
+    solana_sdk::{
+        account::Account, commitment_config::CommitmentConfig, epoch_info::EpochInfo, hash::Hash,
+        message::Message, pubkey::Pubkey, signature::Signature, slot_history::Slot,
+        transaction::Transaction,
+    },
+    solana_transaction_status::UiConfirmedBlock,
+    std::sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+};
+
+#[derive(Debug, Clone, Copy)]
+pub struct HighTpsClientConfig {
+    pub send_batch_size: usize,
+    pub fanout_slots: u64,
+}
+
+type ConnectionCache = BackendConnectionCache<QuicPool, QuicConnectionManager, QuicConfig>;
+pub struct HighTpsClient {
+    leader_tpu_service: LeaderTpuService,
+    exit: Arc<AtomicBool>,
+    rpc_client: Arc<RpcClient>,
+    connection_cache: Arc<ConnectionCache>,
+    config: HighTpsClientConfig,
+}
+
+impl Drop for HighTpsClient {
+    fn drop(&mut self) {
+        self.exit.store(true, Ordering::Relaxed);
+    }
+}
+
+impl HighTpsClient {
+    pub fn new(
+        rpc_client: Arc<RpcClient>,
+        websocket_url: &str,
+        config: HighTpsClientConfig,
+        connection_cache: Arc<ConnectionCache>,
+    ) -> Result<Self> {
+        let exit = Arc::new(AtomicBool::new(false));
+        let create_leader_tpu_service = LeaderTpuService::new(
+            rpc_client.get_inner_client().clone(),
+            websocket_url,
+            Protocol::QUIC,
+            exit.clone(),
+        );
+        let leader_tpu_service = tokio::task::block_in_place(|| {
+            rpc_client.runtime().block_on(create_leader_tpu_service)
+        })?;
+        Ok(Self {
+            leader_tpu_service,
+            exit,
+            rpc_client,
+            connection_cache,
+            config,
+        })
+    }
+}
+
+impl BenchTpsClient for HighTpsClient {
+    fn send_transaction(&self, transaction: Transaction) -> Result<Signature> {
+        self.rpc_client
+            .send_transaction(&transaction)
+            .map_err(|err| err.into())
+    }
+
+    fn send_batch(&self, transactions: Vec<Transaction>) -> Result<()> {
+        let wire_transactions = transactions
+            .into_iter() //.into_par_iter() any effect of this?
+            .map(|tx| bincode::serialize(&tx).expect("transaction should be valid."))
+            .collect::<Vec<_>>();
+        for c in wire_transactions.chunks(self.config.send_batch_size) {
+            let tpu_addresses = self
+                .leader_tpu_service
+                .leader_tpu_sockets(self.config.fanout_slots);
+            for tpu_address in &tpu_addresses {
```

Comment:
only sending to one leader here right?

Reply:
should be to `fanout_slots` leaders. 
In configuration that I specify `1` for this parameter because I think that we don't care how many txs have been timed out but only interested in maximazing TPS. Due to that it makes sense to utilize NIC bandwidth to send as many datagrams to one leader as possible (although it might be that for particular cluster configuration we cannot fully utilize it due to staked congestion control)